---
output: github_document
---

<!-- README.md is generated from README.Rmd. Please edit that file -->

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.path = "man/figures/README-",
  out.width = "100%"
)
```
# GeoTcgaData

<!-- badges: start -->
<!-- badges: end -->

The goal of GeoTcgaData is to deal with RNA-seq, DNA Methylation, and Copy 
number variation data in GEO and TCGA.

## Installation

You can install the released version of GeoTcgaData from [CRAN](https://CRAN.R-project.org) with:

``` r
install.packages("GeoTcgaData")
```
GEO and TCGA provide us with a wealth of data, such as RNA-seq, DNA Methylation, 
and Copy number variation data. It's easy to download data from TCGA using the 
gdc tool, but processing these data into a format suitable for bioinformatics 
analysis requires more work. This R package was developed to handle these data.
## Example

This is a basic example which shows you how to solve a common problem:

```{r example}
library(GeoTcgaData)
## basic example code

#' This function can find the mean value of the gene in each module.
#'
#' param geneExpress a data.frame
#' param module a data.frame
#' param result a string
#'
#' return: a matrix
result <- cal_mean_module(geneExpress,module)

#' Average the expression data of different ids for the same gene in the chip
#' expression profile of GEO or TCGA
#'
#' param file1 a data.frame
#' param k a number
#'
#' return a data.frame

#'
#' examples
aa <- c("Gene Symbol","MARCH1","MARC1","MARCH1","MARCH1","MARCH1")
bb <- c("GSM1629982","2.969058399","4.722410064","8.165514853","8.24243893","8.60815086")
cc <- c("GSM1629982","3.969058399","5.722410064","7.165514853","6.24243893","7.60815086")
file1 <- data.frame(aa=aa,bb=bb,cc=cc)
result <- gene_ave(file1)

#' Get the differentially expressioned genes using DESeq2 package 
#'
#' param profile a data.frame
#'
#' return a data.frame
#'
#' examples
profile2 <- classify_sample(profile)


#' Title
#'
#' param profile2 a result of classify_sample
#'
#' return a matrix
#'
#' examples
profile2 <- classify_sample(profile)
jieguo <- diff_gene(profile2)

#' Merge methylation data downloaded from TCGA
#'
#' param dirr a string for the directory of methylation data download from tcga
#' useing the tools gdc
#' return a matrix
#'
#' examples
mearge_result <- Merge_methy_tcga("direc")

#' Combine clinical information obtained from TCGA and extract survival data
#'
#' param Files_dir1 a dir data
#'
#' return a matrix
#'
#' examples
tcga_cli_deal("your_clinical_directory")

#' Multiple genes symbols may correspond to a same id. Some people think 
#' that the expression value of this id should be 
#' given to each gene, and some people think that the expression value of
#' this id should be deleted. The result of rep1 is to assign the expression
#' of this id to each gene, and rep2 deletes the expression.
#'
#' param file1 input file, a data.frame or a matrixg
#' param string a string,sep of the gene
#'
#' return a data.frame
#'
#' examples
aa <- c("MARCH1 /// MMA","MARC1","MARCH2 /// MARCH3",
        "MARCH3 /// MARCH4","MARCH1")
bb <- c("2.969058399","4.722410064","8.165514853","8.24243893","8.60815086")
cc <- c("3.969058399","5.722410064","7.165514853","6.24243893","7.60815086")
input_fil <- data.frame(aa=aa,bb=bb,cc=cc)
rep1_result <- rep1(input_fil," /// ")
rep1_result <- rep2(input_fil," /// ")

#' Convert  ENSEMBL gene id to gene Symbol in TCGA
#'
#' param profile a data.frame
#'
#' return a data.frame
#'
#' examples
result <- id_conversion(profile)


```



What is special about using `README.Rmd` instead of just `README.md`? You can include R chunks like so:

```{r cars}
summary(cars)
```

You'll still need to render `README.Rmd` regularly, to keep `README.md` up-to-date.

You can also embed plots, for example:

```{r pressure, echo = FALSE}
plot(pressure)
```

In that case, don't forget to commit and push the resulting figure files, so they display on GitHub!
