<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Processing various types of data on GEO and TCGA • GeoTcgaData</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="bootstrap-toc.css">
<script src="bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><meta property="og:title" content="Processing various types of data on GEO and TCGA">
<meta property="og:description" content="Gene Expression Omnibus(GEO) and The Cancer Genome Atlas (TCGA) 
    provide us with a wealth of data, such as RNA-seq, DNA Methylation, SNP
    and Copy number variation data. It's easy to download data from TCGA using the 
    gdc tool, but processing these data into a format suitable for bioinformatics 
    analysis requires more work. This R package was developed to handle these data.">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-home">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">GeoTcgaData</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.1.1.993</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="articles/GeoTcgaData.html">Get started</a>
</li>
<li>
  <a href="reference/index.html">Reference</a>
</li>
<li>
  <a href="news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">
<div class="section level1">
<div class="page-header"><h1 id="geotcgadata">GeoTcgaData<a class="anchor" aria-label="anchor" href="#geotcgadata"></a>
</h1></div>
<p>The goal of GeoTcgaData is to deal with RNA-seq, DNA Methylation, single nucleotide Variation and Copy number variation data in GEO and TCGA.</p>
<div class="section level2">
<h2 id="writing_hand-authors">
<span class="emoji" data-emoji="writing_hand">✍️</span> Authors<a class="anchor" aria-label="anchor" href="#writing_hand-authors"></a>
</h2>
<p>Erqiang Hu</p>
<p>Department of Bioinformatics, School of Basic Medical Sciences, Southern Medical University.</p>
<p><a href="https://cran.r-project.org/package=GeoTcgaData" class="external-link"><img src="http://www.r-pkg.org/badges/version/GeoTcgaData?color=green" alt="CRAN_Status_Badge"></a> <img src="https://cranlogs.r-pkg.org/badges/grand-total/GeoTcgaData?color=green"><img src="https://cranlogs.r-pkg.org/badges/GeoTcgaData?color=green"><img src="https://cranlogs.r-pkg.org/badges/last-week/GeoTcgaData?color=green"><a href="https://www.repostatus.org/#active" class="external-link"><img src="https://www.repostatus.org/badges/latest/active.svg" alt="Project Status: Active – The project has reached a stable, usable state and is being actively developed."></a></p>
</div>
<div class="section level2">
<h2 id="arrow_double_down-installation">
<span class="emoji" data-emoji="arrow_double_down">⏬</span> Installation<a class="anchor" aria-label="anchor" href="#arrow_double_down-installation"></a>
</h2>
<p>Get the released version from CRAN:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"GeoTcgaData"</span><span class="op">)</span></span></code></pre></div>
<p>Or the development version from github:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span><span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html" class="external-link">requireNamespace</a></span><span class="op">(</span><span class="st">"devtools"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"devtools"</span><span class="op">)</span></span>
<span><span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"YuLab-SMU/GeoTcgaData"</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">GeoTcgaData</span><span class="op">)</span></span>
<span><span class="co">#&gt;  =============================================================</span></span>
<span><span class="co">#&gt;  Hello, friend! welcome to use GeoTcgaData!                   </span></span>
<span><span class="co">#&gt;  -------------------------------------------------------------</span></span>
<span><span class="co">#&gt;  Version:1.1.1.993</span></span>
<span><span class="co">#&gt;  =============================================================</span></span></code></pre></div>
<p>GEO and TCGA provide us with a wealth of data, such as RNA-seq, DNA Methylation, single nucleotide Variation and Copy number variation data. It’s easy to download data from TCGA using the gdc tool or <code>TCGAbiolinks</code>, and some software provides organized TCGA data, such as <a href="http://xena.ucsc.edu/" class="external-link">UCSC Xena</a> , <a href="https://cran.r-project.org/package=UCSCXenaTools" class="external-link">UCSCXenaTools</a>，and <a href="http://vip.sangerbox.com/" class="external-link">sangerbox</a>, but processing these data into a format suitable for bioinformatics analysis requires more work. This R package was developed to handle these data.</p>
</div>
<div class="section level2">
<h2 id="example">Example<a class="anchor" aria-label="anchor" href="#example"></a>
</h2>
<p>This is a basic example which shows you how to solve a common problem:</p>
</div>
<div class="section level2">
<h2 id="rna-seq-data-differential-expression-analysis">RNA-seq data differential expression analysis<a class="anchor" aria-label="anchor" href="#rna-seq-data-differential-expression-analysis"></a>
</h2>
<p>It is convenient to use <a href="http://www.bioconductor.org/packages/release/bioc/vignettes/TCGAbiolinks/inst/doc/analysis.html" class="external-link"><code>TCGAbiolinks</code></a> or <a href="https://bioconductor.org/packages/GDCRNATools/" class="external-link"><code>GDCRNATools</code></a> to download and analysis Gene expression data. <code>TCGAbiolinks</code> use <code>edgeR</code> package to do differential expression analysis, while <code>GDCRNATools</code> can implement three most commonly used methods: limma, edgeR , and DESeq2 to identify differentially expressed genes (DEGs).</p>
<p>Alicia Oshlack et. al. claimed that unlike the chip data, the RNA-seq data had one <a href="https://pubmed.ncbi.nlm.nih.gov/20132535/" class="external-link">bias</a>: the larger the transcript length / mean read count , the more likely it was to be identified as a differential gene, <a href="https://pubmed.ncbi.nlm.nih.gov/19371405/" class="external-link">while there was no such trend in the chip data</a>.</p>
<p><img src="reference/figures/Alicia.jpg" width="890"></p>
<p>However, when we use their chip data for difference analysis( using the limma package), we find that chip data has the same trend as RNA-seq data. And we also found this trend in the difference analysis results given by the data <a href="https://genome.cshlp.org/content/18/9/1509.long" class="external-link">authors</a>.</p>
<p><img src="reference/figures/marioni_array_length_pvalue.jpg" width="890"></p>
<p>It is worse noting that <a href="https://pubmed.ncbi.nlm.nih.gov/28545404/" class="external-link">only technical replicate data, which has small gene dispersions, shows this bias</a>. This is because in technical replicate RNA-seq data a long gene has more reads mapping to it compared to a short gene of similar expression, and most of the statistical methods used to detect differential expression have stronger detection ability for genes with more reads. However, we have not deduced why there is such a bias in the current difference analysis algorithms.</p>
<p>Some software, such as <a href="http://www.bioconductor.org/packages/cqn/" class="external-link">CQN</a> , present a <a href="https://pubmed.ncbi.nlm.nih.gov/22285995/" class="external-link">normalization algorithm</a> to correct systematic biases(gene length bias and <a href="https://pubmed.ncbi.nlm.nih.gov/22177264/" class="external-link">GC-content bias</a>. But they did not provide sufficient evidence to prove that the correction is effective. We use the <a href="https://pubmed.ncbi.nlm.nih.gov/19371405/" class="external-link">Marioni dataset</a> to verify the correction effect of CQN and find that there is still a deviation after correction:</p>
<p><img src="reference/figures/marioni_bin_de.jpg" width="890"></p>
<p><a href="http://bioconductor.org/packages/goseq/" class="external-link">GOseq</a> based on <a href="https://en.wikipedia.org/wiki/Wallenius%27_noncentral_hypergeometric_distribution" class="external-link">Wallenius’ noncentral hypergeometric distribution</a> can effectively correct the gene length deviation in enrichment analysis. However, the current RNA-seq data often have no gene length bias, but only the expression amount(read count) bias, GOseq may overcorrect these data, correcting originally unbiased data into reverse bias. <img src="reference/figures/goseq.jpg" width="890"></p>
<p>GOseq also fails to correct for expression bias, therefore, read count bias correction is still a challenge for us. <img src="reference/figures/goseq_marioni.jpg" width="890"></p>
<p>use <code>TCGAbiolinks</code> to download TCGA data</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># download RNA-seq data</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/BioinformaticsFMRP/TCGAbiolinks" class="external-link">TCGAbiolinks</a></span><span class="op">)</span></span>
<span>                </span>
<span><span class="va">query</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/TCGAbiolinks/man/GDCquery.html" class="external-link">GDCquery</a></span><span class="op">(</span>project <span class="op">=</span> <span class="st">"TCGA-ACC"</span>,</span>
<span>                  data.category <span class="op">=</span> <span class="st">"Transcriptome Profiling"</span>,</span>
<span>                  data.type <span class="op">=</span> <span class="st">"Gene Expression Quantification"</span>, </span>
<span>                  workflow.type <span class="op">=</span> <span class="st">"STAR - Counts"</span><span class="op">)</span></span>
<span>                  </span>
<span><span class="fu"><a href="https://rdrr.io/pkg/TCGAbiolinks/man/GDCdownload.html" class="external-link">GDCdownload</a></span><span class="op">(</span><span class="va">query</span>, method <span class="op">=</span> <span class="st">"api"</span>, files.per.chunk <span class="op">=</span> <span class="fl">3</span>, </span>
<span>    directory <span class="op">=</span> <span class="va">Your_Path</span><span class="op">)</span></span>
<span></span>
<span><span class="va">dataRNA</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/TCGAbiolinks/man/GDCprepare.html" class="external-link">GDCprepare</a></span><span class="op">(</span>query <span class="op">=</span> <span class="va">query</span>, directory <span class="op">=</span> <span class="va">Your_Path</span>,</span>
<span>                      save <span class="op">=</span> <span class="cn">TRUE</span>, save.filename <span class="op">=</span> <span class="st">"dataRNA.RData"</span><span class="op">)</span></span>
<span><span class="co">## get raw count matrix                         </span></span>
<span><span class="va">dataPrep</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/TCGAbiolinks/man/TCGAanalyze_Preprocessing.html" class="external-link">TCGAanalyze_Preprocessing</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">dataRNA</span>,</span>
<span>                                      cor.cut <span class="op">=</span> <span class="fl">0.6</span>,</span>
<span>                                      datatype <span class="op">=</span> <span class="st">"STAR - Counts"</span><span class="op">)</span></span></code></pre></div>
<p>Use <code>differential_RNA</code> to do difference analysis. We provide the data of human gene length and GC content in <code>gene_cov</code>.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">group</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/AnnotationDbi/man/Bimap-envirAPI.html" class="external-link">sample</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"grp1"</span>, <span class="st">"grp2"</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">dataPrep</span><span class="op">)</span>, replace <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">cqn</span><span class="op">)</span> <span class="co"># To avoid reporting errors: there is no function "rq"</span></span>
<span><span class="co">## get gene length and GC content</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">org.Hs.eg.db</span><span class="op">)</span></span>
<span><span class="va">genes_bitr</span> <span class="op">&lt;-</span> <span class="fu">bitr</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">gene_cov</span><span class="op">)</span>, fromType <span class="op">=</span> <span class="st">"ENTREZID"</span>, toType <span class="op">=</span> <span class="st">"ENSEMBL"</span>, </span>
<span>         OrgDb <span class="op">=</span> <span class="va">org.Hs.eg.db</span>, drop <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">genes_bitr</span> <span class="op">&lt;-</span> <span class="va">genes_bitr</span><span class="op">[</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/duplicated.html" class="external-link">duplicated</a></span><span class="op">(</span><span class="va">genes_bitr</span><span class="op">[</span>,<span class="fl">2</span><span class="op">]</span><span class="op">)</span>, <span class="op">]</span></span>
<span><span class="va">gene_cov2</span> <span class="op">&lt;-</span> <span class="va">gene_cov</span><span class="op">[</span><span class="va">genes_bitr</span><span class="op">$</span><span class="va">ENTREZID</span>, <span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">gene_cov2</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">genes_bitr</span><span class="op">$</span><span class="va">ENSEMBL</span></span>
<span><span class="va">genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sets.html" class="external-link">intersect</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">dataPrep</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">gene_cov2</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">dataPrep</span> <span class="op">&lt;-</span> <span class="va">dataPrep</span><span class="op">[</span><span class="va">genes</span>, <span class="op">]</span></span>
<span><span class="va">geneLength</span> <span class="op">&lt;-</span> <span class="va">gene_cov2</span><span class="op">[</span><span class="va">genes</span>, <span class="st">"length"</span><span class="op">]</span></span>
<span><span class="va">gccontent</span> <span class="op">&lt;-</span> <span class="va">gene_cov2</span><span class="op">[</span><span class="va">genes</span>, <span class="st">"GC"</span><span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">geneLength</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">gccontent</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">genes</span></span>
<span><span class="co">##  Difference analysis</span></span>
<span><span class="va">DEGAll</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/differential_RNA.html">differential_RNA</a></span><span class="op">(</span>counts <span class="op">=</span> <span class="va">dataPrep</span>, group <span class="op">=</span> <span class="va">group</span>, </span>
<span>                   geneLength <span class="op">=</span> <span class="va">geneLength</span>, gccontent <span class="op">=</span> <span class="va">gccontent</span><span class="op">)</span></span></code></pre></div>
<p>Use <code>clusterProfiler</code> to do enrichment analytics:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">diffGenes</span> <span class="op">&lt;-</span> <span class="va">DEGAll</span><span class="op">$</span><span class="va">logFC</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">diffGenes</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">DEGAll</span><span class="op">)</span></span>
<span><span class="va">diffGenes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sort.html" class="external-link">sort</a></span><span class="op">(</span><span class="va">diffGenes</span>, decreasing <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://yulab-smu.top/biomedical-knowledge-mining-book/" class="external-link">clusterProfiler</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://yulab-smu.top/biomedical-knowledge-mining-book/" class="external-link">enrichplot</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">org.Hs.eg.db</span><span class="op">)</span></span>
<span><span class="va">gsego</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/clusterProfiler/man/gseGO.html" class="external-link">gseGO</a></span><span class="op">(</span>gene <span class="op">=</span> <span class="va">diffGenes</span>, OrgDb <span class="op">=</span> <span class="va">org.Hs.eg.db</span>, keyType <span class="op">=</span> <span class="st">"ENSEMBL"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/enrichplot/man/dotplot.html" class="external-link">dotplot</a></span><span class="op">(</span><span class="va">gsego</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="dna-methylation-data-integration">DNA Methylation data integration<a class="anchor" aria-label="anchor" href="#dna-methylation-data-integration"></a>
</h2>
<p>use <code>TCGAbiolinks</code> to download TCGA data.</p>
<p>The codes may need to be modified if <code>TCGAbiolinks</code> updates. So please read its <a href="https://www.bioconductor.org/packages/release/bioc/html/TCGAbiolinks.html" class="external-link">documents</a>.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/BioinformaticsFMRP/TCGAbiolinks" class="external-link">TCGAbiolinks</a></span><span class="op">)</span></span>
<span><span class="va">query</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/TCGAbiolinks/man/GDCquery.html" class="external-link">GDCquery</a></span><span class="op">(</span>project <span class="op">=</span> <span class="st">"TCGA-ACC"</span>,</span>
<span>                  data.category <span class="op">=</span> <span class="st">"DNA Methylation"</span>,</span>
<span>                  data.type <span class="op">=</span> <span class="st">"Methylation Beta Value"</span>,</span>
<span>                  platform <span class="op">=</span> <span class="st">"Illumina Human Methylation 450"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/TCGAbiolinks/man/GDCdownload.html" class="external-link">GDCdownload</a></span><span class="op">(</span><span class="va">query</span>, method <span class="op">=</span> <span class="st">"api"</span>, files.per.chunk <span class="op">=</span> <span class="fl">5</span>, directory <span class="op">=</span> <span class="va">Your_Path</span><span class="op">)</span></span></code></pre></div>
<p>The function <code>Merge_methy_tcga</code> could Merge methylation data downloaded from TCGA official website or TCGAbiolinks. This makes it easier to extract differentially methylated genes in the downstream analysis. For example:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">merge_result</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/Merge_methy_tcga.html">Merge_methy_tcga</a></span><span class="op">(</span><span class="va">Your_Path_to_DNA_Methylation_data</span><span class="op">)</span></span></code></pre></div>
<p>Then use <code>ChAMP</code> package to do difference analysis.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html" class="external-link">requireNamespace</a></span><span class="op">(</span><span class="st">"ChAMP"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">"ChAMP"</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">ChAMP</span><span class="op">)</span> <span class="co"># To avoid reporting errors</span></span>
<span><span class="va">differential_gene</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/methyDiff.html">methyDiff</a></span><span class="op">(</span>cpgData <span class="op">=</span> <span class="va">merge_result</span>, sampleGroup <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/AnnotationDbi/man/Bimap-envirAPI.html" class="external-link">sample</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"C"</span>,<span class="st">"T"</span><span class="op">)</span>, </span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">merge_result</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span>, replace <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><strong>Note:</strong> <code>ChAMP</code>has a large number of dependent packages. If you cannot install it successfully, you can download each dependent package separately(Source or Binary) and install it locally.</p>
<p>If your methylation data was downloaded from <a href="http://xena.ucsc.edu/" class="external-link">UCSC Xena</a>, you can use <code>methydifferential_ucsc</code> to get differential genes.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">methy_file</span> <span class="op">&lt;-</span> <span class="st">"TCGA.THCA.sampleMap_HumanMethylation450.gz"</span></span>
<span><span class="va">methy</span> <span class="op">&lt;-</span> <span class="fu">fread</span><span class="op">(</span><span class="va">methy_file</span>, sep <span class="op">=</span> <span class="st">"\t"</span>, header <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">ChAMP</span><span class="op">)</span></span>
<span><span class="va">myImport</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ChAMP/man/champ.import.html" class="external-link">champ.import</a></span><span class="op">(</span>directory<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>,package<span class="op">=</span><span class="st">"ChAMPdata"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">myfilter</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ChAMP/man/champ.filter.html" class="external-link">champ.filter</a></span><span class="op">(</span>beta<span class="op">=</span><span class="va">myImport</span><span class="op">$</span><span class="va">beta</span>,pd<span class="op">=</span><span class="va">myImport</span><span class="op">$</span><span class="va">pd</span>,detP<span class="op">=</span><span class="va">myImport</span><span class="op">$</span><span class="va">detP</span>,beadcount<span class="op">=</span><span class="va">myImport</span><span class="op">$</span><span class="va">beadcount</span><span class="op">)</span></span>
<span><span class="va">cpg_gene</span> <span class="op">&lt;-</span> <span class="va">hm450.manifest.hg19</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"probeID"</span>, <span class="st">"gene_HGNC"</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="co">## or use IlluminaHumanMethylation450kanno.ilmn12.hg19 to get annotation data</span></span>
<span><span class="co"># library(IlluminaHumanMethylation450kanno.ilmn12.hg19)</span></span>
<span><span class="co"># ann &lt;- getAnnotation(IlluminaHumanMethylation450kanno.ilmn12.hg19)</span></span>
<span><span class="co"># class(ann) &lt;- "data.frame"</span></span>
<span><span class="co"># cpg_gene &lt;- ann[,c("Name", "UCSC_RefGene_Name", "UCSC_RefGene_Group")]</span></span>
<span></span>
<span><span class="va">methy_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/methydifferential_ucsc.html">methydifferential_ucsc</a></span><span class="op">(</span><span class="va">methy</span>, <span class="va">cpg_gene</span><span class="op">)</span></span></code></pre></div>
<p>We provide three models to get methylation difference genes:</p>
<p>if model = “cpg”, step1: calculate difference cpgs; step2: calculate difference genes;</p>
<p>if model = “gene”, step1: calculate the methylation level of genes; step2: calculate difference genes.</p>
<p>We find that only model = “gene” has no deviation of CpG number.</p>
<p><img src="reference/figures/methy.jpg" width="890"></p>
<p>Use <code>clusterProfiler</code> to do enrichment analytics:</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">differential_gene</span><span class="op">$</span><span class="va">p.adj</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/p.adjust.html" class="external-link">p.adjust</a></span><span class="op">(</span><span class="va">differential_gene</span><span class="op">$</span><span class="va">pvalue</span><span class="op">)</span></span>
<span><span class="va">genes</span> <span class="op">&lt;-</span> <span class="va">differential_gene</span><span class="op">[</span><span class="va">differential_gene</span><span class="op">$</span><span class="va">p.adj</span> <span class="op">&lt;</span> <span class="fl">0.05</span>, <span class="st">"gene"</span><span class="op">]</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://yulab-smu.top/biomedical-knowledge-mining-book/" class="external-link">clusterProfiler</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://yulab-smu.top/biomedical-knowledge-mining-book/" class="external-link">enrichplot</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">org.Hs.eg.db</span><span class="op">)</span></span>
<span><span class="va">ego</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/clusterProfiler/man/enrichGO.html" class="external-link">enrichGO</a></span><span class="op">(</span>gene <span class="op">=</span> <span class="va">genes</span>, OrgDb <span class="op">=</span> <span class="va">org.Hs.eg.db</span>, keyType <span class="op">=</span> <span class="st">"SYMBOL"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/enrichplot/man/dotplot.html" class="external-link">dotplot</a></span><span class="op">(</span><span class="va">ego</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="copy-number-variation-data-integration-and-differential-gene-extraction">Copy number variation data integration and differential gene extraction<a class="anchor" aria-label="anchor" href="#copy-number-variation-data-integration-and-differential-gene-extraction"></a>
</h2>
<p>use TCGAbiolinks to download TCGA data(Gene Level Copy Number Scores)</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/BioinformaticsFMRP/TCGAbiolinks" class="external-link">TCGAbiolinks</a></span><span class="op">)</span></span>
<span><span class="va">query</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/TCGAbiolinks/man/GDCquery.html" class="external-link">GDCquery</a></span><span class="op">(</span>project <span class="op">=</span> <span class="st">"TCGA-LGG"</span>,</span>
<span>                  data.category <span class="op">=</span> <span class="st">"Copy Number Variation"</span>,</span>
<span>                  data.type <span class="op">=</span> <span class="st">"Gene Level Copy Number Scores"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/TCGAbiolinks/man/GDCdownload.html" class="external-link">GDCdownload</a></span><span class="op">(</span><span class="va">query</span>, method <span class="op">=</span> <span class="st">"api"</span>, files.per.chunk <span class="op">=</span> <span class="fl">5</span>, directory <span class="op">=</span> <span class="va">Your_Path</span><span class="op">)</span></span>
<span></span>
<span><span class="va">data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/TCGAbiolinks/man/GDCprepare.html" class="external-link">GDCprepare</a></span><span class="op">(</span>query <span class="op">=</span> <span class="va">query</span>, </span>
<span>                   directory <span class="op">=</span>  <span class="va">Your_Path</span><span class="op">)</span> </span></code></pre></div>
<p>Do difference analysis of gene level copy number variation data using <code>differential_CNV</code></p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/class.html" class="external-link">class</a></span><span class="op">(</span><span class="va">data</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">"data.frame"</span></span>
<span><span class="va">cnvData</span> <span class="op">&lt;-</span> <span class="va">data</span><span class="op">[</span>, <span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">2</span>,<span class="fl">3</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">cnvData</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">data</span><span class="op">[</span>, <span class="fl">1</span><span class="op">]</span></span>
<span><span class="va">sampleGroup</span>  <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/AnnotationDbi/man/Bimap-envirAPI.html" class="external-link">sample</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"A"</span>,<span class="st">"B"</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">cnvData</span><span class="op">)</span>, replace <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">diffCnv</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/differential_CNV.html">differential_CNV</a></span><span class="op">(</span><span class="va">cnvData</span>, <span class="va">sampleGroup</span><span class="op">)</span></span></code></pre></div>
<p>Use <code>clusterProfiler</code> to do enrichment analytics:</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pvalues</span> <span class="op">&lt;-</span> <span class="va">diffCnv</span><span class="op">$</span><span class="va">pvalue</span> <span class="op">*</span> <span class="fu"><a href="https://rdrr.io/r/base/sign.html" class="external-link">sign</a></span><span class="op">(</span><span class="va">diffCnv</span><span class="op">$</span><span class="va">odds</span><span class="op">)</span></span>
<span><span class="va">genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">diffCnv</span><span class="op">)</span><span class="op">[</span><span class="va">diffCnv</span><span class="op">$</span><span class="va">pvalue</span> <span class="op">&lt;</span> <span class="fl">0.05</span><span class="op">]</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://yulab-smu.top/biomedical-knowledge-mining-book/" class="external-link">clusterProfiler</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://yulab-smu.top/biomedical-knowledge-mining-book/" class="external-link">enrichplot</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">org.Hs.eg.db</span><span class="op">)</span></span>
<span><span class="va">ego</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/clusterProfiler/man/enrichGO.html" class="external-link">enrichGO</a></span><span class="op">(</span>gene <span class="op">=</span> <span class="va">genes</span>, OrgDb <span class="op">=</span> <span class="va">org.Hs.eg.db</span>, keyType <span class="op">=</span> <span class="st">"ENSEMBL"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/enrichplot/man/dotplot.html" class="external-link">dotplot</a></span><span class="op">(</span><span class="va">ego</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="difference-analysis-of-single-nucleotide-variation-data">Difference analysis of single nucleotide Variation data<a class="anchor" aria-label="anchor" href="#difference-analysis-of-single-nucleotide-variation-data"></a>
</h2>
<p>Use TCGAbiolinks to download TCGA data</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/BioinformaticsFMRP/TCGAbiolinks" class="external-link">TCGAbiolinks</a></span><span class="op">)</span></span>
<span><span class="va">query</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/TCGAbiolinks/man/GDCquery.html" class="external-link">GDCquery</a></span><span class="op">(</span>project <span class="op">=</span> <span class="st">"TCGA-ACC"</span>,</span>
<span>                  data.category <span class="op">=</span> <span class="st">"Simple Nucleotide Variation"</span>,</span>
<span>                  data.type <span class="op">=</span> <span class="st">"Masked Somatic Mutation"</span>,</span>
<span>                  workflow.type <span class="op">=</span> <span class="st">"MuSE Variant Aggregation and Masking"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/TCGAbiolinks/man/GDCdownload.html" class="external-link">GDCdownload</a></span><span class="op">(</span><span class="va">query</span>, method <span class="op">=</span> <span class="st">"api"</span>, files.per.chunk <span class="op">=</span> <span class="fl">5</span>, directory <span class="op">=</span> <span class="va">Your_Path</span><span class="op">)</span></span>
<span></span>
<span><span class="va">data_snp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/TCGAbiolinks/man/GDCprepare.html" class="external-link">GDCprepare</a></span><span class="op">(</span>query <span class="op">=</span> <span class="va">query</span>, </span>
<span>                   directory <span class="op">=</span>  <span class="va">Your_Path</span><span class="op">)</span> </span></code></pre></div>
<p>Use <code>differential_SNP_tcga</code> to do difference analysis</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">samples</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">data_snp</span><span class="op">$</span><span class="va">Tumor_Sample_Barcode</span><span class="op">)</span></span>
<span><span class="va">sampleType</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/AnnotationDbi/man/Bimap-envirAPI.html" class="external-link">sample</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"A"</span>,<span class="st">"B"</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">samples</span><span class="op">)</span>, replace <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">sampleType</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">samples</span></span>
<span><span class="va">pvalue</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/differential_SNP_tcga.html">differential_SNP_tcga</a></span><span class="op">(</span>snpData <span class="op">=</span> <span class="va">data_snp</span>, sampleType <span class="op">=</span> <span class="va">sampleType</span><span class="op">)</span></span>
<span><span class="co"># merge pvalue</span></span></code></pre></div>
<p>Use <code>clusterProfiler</code> to do enrichment analysis</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pvalue2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sort.html" class="external-link">sort</a></span><span class="op">(</span><span class="va">pvalue</span>, decreasing <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://yulab-smu.top/biomedical-knowledge-mining-book/" class="external-link">clusterProfiler</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://yulab-smu.top/biomedical-knowledge-mining-book/" class="external-link">enrichplot</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">org.Hs.eg.db</span><span class="op">)</span></span>
<span><span class="va">gsego</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/clusterProfiler/man/gseGO.html" class="external-link">gseGO</a></span><span class="op">(</span><span class="va">pvalue2</span>, OrgDb <span class="op">=</span> <span class="va">org.Hs.eg.db</span>, keyType <span class="op">=</span> <span class="st">"SYMBOL"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/enrichplot/man/dotplot.html" class="external-link">dotplot</a></span><span class="op">(</span><span class="va">gsego</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="geo-chip-data-processing">GEO chip data processing<a class="anchor" aria-label="anchor" href="#geo-chip-data-processing"></a>
</h2>
<p>The function <code>gene_ave</code> could average the expression data of different ids for the same gene in the GEO chip data. For example:</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">aa</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"MARCH1"</span>,<span class="st">"MARC1"</span>,<span class="st">"MARCH1"</span>,<span class="st">"MARCH1"</span>,<span class="st">"MARCH1"</span><span class="op">)</span></span>
<span><span class="va">bb</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">2.969058399</span>,<span class="fl">4.722410064</span>,<span class="fl">8.165514853</span>,<span class="fl">8.24243893</span>,<span class="fl">8.60815086</span><span class="op">)</span></span>
<span><span class="va">cc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">3.969058399</span>,<span class="fl">5.722410064</span>,<span class="fl">7.165514853</span>,<span class="fl">6.24243893</span>,<span class="fl">7.60815086</span><span class="op">)</span></span>
<span><span class="va">file_gene_ave</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>aa<span class="op">=</span><span class="va">aa</span>,bb<span class="op">=</span><span class="va">bb</span>,cc<span class="op">=</span><span class="va">cc</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">file_gene_ave</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Gene"</span>, <span class="st">"GSM1629982"</span>, <span class="st">"GSM1629983"</span><span class="op">)</span></span>
<span><span class="va">result</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/gene_ave.html">gene_ave</a></span><span class="op">(</span><span class="va">file_gene_ave</span>, <span class="fl">1</span><span class="op">)</span></span></code></pre></div>
<p>Multiple genes symbols may correspond to a same chip id. The result of function <code>repAssign</code> is to assign the expression of this id to each gene, and function <code>repRemove</code> deletes the expression. For example:</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">aa</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"MARCH1 /// MMA"</span>,<span class="st">"MARC1"</span>,<span class="st">"MARCH2 /// MARCH3"</span>,<span class="st">"MARCH3 /// MARCH4"</span>,<span class="st">"MARCH1"</span><span class="op">)</span></span>
<span><span class="va">bb</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"2.969058399"</span>,<span class="st">"4.722410064"</span>,<span class="st">"8.165514853"</span>,<span class="st">"8.24243893"</span>,<span class="st">"8.60815086"</span><span class="op">)</span></span>
<span><span class="va">cc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"3.969058399"</span>,<span class="st">"5.722410064"</span>,<span class="st">"7.165514853"</span>,<span class="st">"6.24243893"</span>,<span class="st">"7.60815086"</span><span class="op">)</span></span>
<span><span class="va">input_file</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>aa<span class="op">=</span><span class="va">aa</span>,bb<span class="op">=</span><span class="va">bb</span>,cc<span class="op">=</span><span class="va">cc</span><span class="op">)</span></span>
<span></span>
<span><span class="va">repAssign_result</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/repAssign.html">repAssign</a></span><span class="op">(</span><span class="va">input_file</span>,<span class="st">" /// "</span><span class="op">)</span></span>
<span><span class="va">repRemove_result</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/repRemove.html">repRemove</a></span><span class="op">(</span><span class="va">input_file</span>,<span class="st">" /// "</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="other-downstream-analyses">Other downstream analyses<a class="anchor" aria-label="anchor" href="#other-downstream-analyses"></a>
</h2>
<ol style="list-style-type: decimal"><li>The function <code>id_conversion</code> could convert gene id from one of <code>symbol</code>, <code>RefSeq_ID</code>, <code>Ensembl_ID</code>, <code>NCBI_Gene_ID</code>, <code>UCSC_ID</code>, and <code>UniProt_ID</code> , etc. to another. Use <code><a href="reference/id_ava.html">id_ava()</a></code> to get all the convertible ids. For example:</li></ol>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="reference/id_conversion.html">id_conversion</a></span><span class="op">(</span><span class="st">"symbol"</span>, <span class="st">"ensembl_gene_id"</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"A2ML1"</span>, <span class="st">"A2ML1-AS1"</span>, <span class="st">"A4GALT"</span>, <span class="st">"A12M1"</span>, <span class="st">"AAAS"</span><span class="op">)</span><span class="op">)</span> </span>
<span><span class="co">#&gt; 80% were successfully converted.</span></span>
<span><span class="co">#&gt;        from              to</span></span>
<span><span class="co">#&gt; 1     A2ML1 ENSG00000166535</span></span>
<span><span class="co">#&gt; 2 A2ML1-AS1 ENSG00000256661</span></span>
<span><span class="co">#&gt; 3    A4GALT ENSG00000128274</span></span>
<span><span class="co">#&gt; 4      AAAS ENSG00000094914</span></span></code></pre></div>
<p>When the user converts the Ensembl ID to other ids, the version number needs to be removed. For example, “ENSG00000186092.4” doesn’t work, you need to change it to “ENSG00000186092”.</p>
<p>Especially, the function id_conversion could convert ENSEMBL gene id to gene Symbol in TCGA. For example:</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">result</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/id_conversion.html">id_conversion</a></span><span class="op">(</span><span class="va">profile</span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; 'select()' returned 1:1 mapping between keys and columns</span></span>
<span><span class="co">#&gt; Warning in clusterProfiler::bitr(rownames(profiles), fromType = "ENSEMBL", :</span></span>
<span><span class="co">#&gt; 33.33% of input gene IDs are fail to map...</span></span></code></pre></div>
<p>The parameter profile is a data.frame or matrix of gene expression data in TCGA.</p>
<p><strong>Note:</strong> In previous versions(&lt; 1.0.0) the <code>id_conversion</code> and <code>id_conversion</code> used HGNC data to convert human gene id. In future versions, we will use <code><a href="https://rdrr.io/pkg/clusterProfiler/man/bitr.html" class="external-link">clusterProfiler::bitr</a></code> for ID conversion.</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://yulab-smu.top/biomedical-knowledge-mining-book/" class="external-link">clusterProfiler</a></span><span class="op">)</span></span>
<span><span class="co">#&gt; clusterProfiler v4.5.2.001  For help: https://yulab-smu.top/biomedical-knowledge-mining-book/</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; If you use clusterProfiler in published research, please cite:</span></span>
<span><span class="co">#&gt; T Wu, E Hu, S Xu, M Chen, P Guo, Z Dai, T Feng, L Zhou, W Tang, L Zhan, X Fu, S Liu, X Bo, and G Yu. clusterProfiler 4.0: A universal enrichment tool for interpreting omics data. The Innovation. 2021, 2(3):100141</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; 载入程辑包：'clusterProfiler'</span></span>
<span><span class="co">#&gt; The following object is masked from 'package:stats':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     filter</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">org.Hs.eg.db</span><span class="op">)</span></span>
<span><span class="co">#&gt; 载入需要的程辑包：AnnotationDbi</span></span>
<span><span class="co">#&gt; 载入需要的程辑包：stats4</span></span>
<span><span class="co">#&gt; 载入需要的程辑包：BiocGenerics</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; 载入程辑包：'BiocGenerics'</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:stats':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     IQR, mad, sd, var, xtabs</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:base':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     anyDuplicated, append, as.data.frame, basename, cbind, colnames,</span></span>
<span><span class="co">#&gt;     dirname, do.call, duplicated, eval, evalq, Filter, Find, get, grep,</span></span>
<span><span class="co">#&gt;     grepl, intersect, is.unsorted, lapply, Map, mapply, match, mget,</span></span>
<span><span class="co">#&gt;     order, paste, pmax, pmax.int, pmin, pmin.int, Position, rank,</span></span>
<span><span class="co">#&gt;     rbind, Reduce, rownames, sapply, setdiff, sort, table, tapply,</span></span>
<span><span class="co">#&gt;     union, unique, unsplit, which.max, which.min</span></span>
<span><span class="co">#&gt; 载入需要的程辑包：Biobase</span></span>
<span><span class="co">#&gt; Welcome to Bioconductor</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     Vignettes contain introductory material; view with</span></span>
<span><span class="co">#&gt;     'browseVignettes()'. To cite Bioconductor, see</span></span>
<span><span class="co">#&gt;     'citation("Biobase")', and for packages 'citation("pkgname")'.</span></span>
<span><span class="co">#&gt; 载入需要的程辑包：IRanges</span></span>
<span><span class="co">#&gt; 载入需要的程辑包：S4Vectors</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; 载入程辑包：'S4Vectors'</span></span>
<span><span class="co">#&gt; The following object is masked from 'package:clusterProfiler':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     rename</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:base':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     expand.grid, I, unname</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; 载入程辑包：'IRanges'</span></span>
<span><span class="co">#&gt; The following object is masked from 'package:clusterProfiler':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     slice</span></span>
<span><span class="co">#&gt; The following object is masked from 'package:grDevices':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     windows</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; 载入程辑包：'AnnotationDbi'</span></span>
<span><span class="co">#&gt; The following object is masked from 'package:clusterProfiler':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     select</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/clusterProfiler/man/bitr.html" class="external-link">bitr</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"A2ML1"</span>, <span class="st">"A2ML1-AS1"</span>, <span class="st">"A4GALT"</span>, <span class="st">"A12M1"</span>, <span class="st">"AAAS"</span><span class="op">)</span>, fromType <span class="op">=</span> <span class="st">"SYMBOL"</span>, </span>
<span>     toType <span class="op">=</span> <span class="st">"ENSEMBL"</span>, OrgDb <span class="op">=</span> <span class="va">org.Hs.eg.db</span>, drop <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="co">#&gt; 'select()' returned 1:1 mapping between keys and columns</span></span>
<span><span class="co">#&gt; Warning in bitr(c("A2ML1", "A2ML1-AS1", "A4GALT", "A12M1", "AAAS"), fromType =</span></span>
<span><span class="co">#&gt; "SYMBOL", : 40% of input gene IDs are fail to map...</span></span>
<span><span class="co">#&gt;      SYMBOL         ENSEMBL</span></span>
<span><span class="co">#&gt; 1     A2ML1 ENSG00000166535</span></span>
<span><span class="co">#&gt; 2 A2ML1-AS1            &lt;NA&gt;</span></span>
<span><span class="co">#&gt; 3    A4GALT ENSG00000128274</span></span>
<span><span class="co">#&gt; 4     A12M1            &lt;NA&gt;</span></span>
<span><span class="co">#&gt; 5      AAAS ENSG00000094914</span></span></code></pre></div>
<ol start="2" style="list-style-type: decimal"><li>The function <code>countToFpkm</code> and <code>countToTpm</code> could convert count data to FPKM or TPM data.</li></ol>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">lung_squ_count2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">2</span>, <span class="fl">3</span>, <span class="fl">4</span>, <span class="fl">5</span>, <span class="fl">6</span>, <span class="fl">7</span>, <span class="fl">8</span>, <span class="fl">9</span><span class="op">)</span>, ncol <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">lung_squ_count2</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"DISC1"</span>, <span class="st">"TCOF1"</span>, <span class="st">"SPPL3"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">lung_squ_count2</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"sample1"</span>, <span class="st">"sample2"</span>, <span class="st">"sample3"</span><span class="op">)</span></span>
<span><span class="va">result</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/countToFpkm.html">countToFpkm</a></span><span class="op">(</span><span class="va">lung_squ_count2</span>,</span>
<span>  keyType <span class="op">=</span> <span class="st">"SYMBOL"</span>,</span>
<span>  gene_cov <span class="op">=</span> <span class="va">gene_cov</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt; 'select()' returned 1:1 mapping between keys and columns</span></span>
<span><span class="co">#&gt; Warning in clusterProfiler::bitr(rownames(gene_cov), fromType = "ENTREZID", :</span></span>
<span><span class="co">#&gt; 0.05% of input gene IDs are fail to map...</span></span>
<span><span class="va">result</span></span>
<span><span class="co">#&gt;        sample1  sample2  sample3</span></span>
<span><span class="co">#&gt; DISC1 11449.25 18318.79 20036.18</span></span>
<span><span class="co">#&gt; TCOF1 29140.08 29140.08 29140.08</span></span>
<span><span class="co">#&gt; SPPL3 69473.39 55578.71 52105.04</span></span></code></pre></div>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">lung_squ_count2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">2</span>, <span class="fl">3</span>, <span class="fl">4</span>, <span class="fl">5</span>, <span class="fl">6</span>, <span class="fl">7</span>, <span class="fl">8</span>, <span class="fl">9</span><span class="op">)</span>, ncol <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">lung_squ_count2</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"DISC1"</span>, <span class="st">"TCOF1"</span>, <span class="st">"SPPL3"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">lung_squ_count2</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"sample1"</span>, <span class="st">"sample2"</span>, <span class="st">"sample3"</span><span class="op">)</span></span>
<span><span class="va">result</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/countToTpm.html">countToTpm</a></span><span class="op">(</span><span class="va">lung_squ_count2</span>,</span>
<span>  keyType <span class="op">=</span> <span class="st">"SYMBOL"</span>,</span>
<span>  gene_cov <span class="op">=</span> <span class="va">gene_cov</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt; 'select()' returned 1:1 mapping between keys and columns</span></span>
<span><span class="co">#&gt; Warning in clusterProfiler::bitr(rownames(gene_cov), fromType = "ENTREZID", :</span></span>
<span><span class="co">#&gt; 0.05% of input gene IDs are fail to map...</span></span>
<span><span class="va">result</span></span>
<span><span class="co">#&gt;        sample1  sample2  sample3</span></span>
<span><span class="co">#&gt; DISC1 104024.7 177787.5 197827.0</span></span>
<span><span class="co">#&gt; TCOF1 264758.8 282810.2 287714.3</span></span>
<span><span class="co">#&gt; SPPL3 631216.4 539402.3 514458.7</span></span></code></pre></div>
<ol start="3" style="list-style-type: decimal"><li>The function <code>tcga_cli_deal</code> could combine clinical information obtained from TCGA and extract survival data. For example:</li></ol>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">tcga_cli</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/tcga_cli_deal.html">tcga_cli_deal</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="st">"extdata"</span>,<span class="st">"tcga_cli"</span><span class="op">)</span>,package<span class="op">=</span><span class="st">"GeoTcgaData"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><strong>Note:</strong> Now the combined clinical data can be downloaded directly from <a href="http://www.bioconductor.org/packages/release/bioc/vignettes/TCGAbiolinks/inst/doc/clinical.html" class="external-link">TCGAbiolinks</a>.</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/BioinformaticsFMRP/TCGAbiolinks" class="external-link">TCGAbiolinks</a></span><span class="op">)</span></span>
<span><span class="co">## get BCR Biotab data</span></span>
<span><span class="va">query</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/TCGAbiolinks/man/GDCquery.html" class="external-link">GDCquery</a></span><span class="op">(</span>project <span class="op">=</span> <span class="st">"TCGA-ACC"</span>, </span>
<span>                  data.category <span class="op">=</span> <span class="st">"Clinical"</span>,</span>
<span>                  data.type <span class="op">=</span> <span class="st">"Clinical Supplement"</span>, </span>
<span>                  data.format <span class="op">=</span> <span class="st">"BCR Biotab"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/TCGAbiolinks/man/GDCdownload.html" class="external-link">GDCdownload</a></span><span class="op">(</span><span class="va">query</span><span class="op">)</span></span>
<span><span class="va">clinical.BCRtab.all</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/TCGAbiolinks/man/GDCprepare.html" class="external-link">GDCprepare</a></span><span class="op">(</span><span class="va">query</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">clinical.BCRtab.all</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## get indexed data</span></span>
<span><span class="va">clinical</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/TCGAbiolinks/man/GDCquery_clinic.html" class="external-link">GDCquery_clinic</a></span><span class="op">(</span>project <span class="op">=</span> <span class="st">"TCGA-ACC"</span>, type <span class="op">=</span> <span class="st">"clinical"</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <div class="links">
<h2 data-toc-skip>Links</h2>
<ul class="list-unstyled">
<li><a href="https://cloud.r-project.org/package=GeoTcgaData" class="external-link">View on CRAN</a></li>
</ul>
</div>

<div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li><a href="https://www.r-project.org/Licenses/GPL-3" class="external-link">GPL-3</a></li>
</ul>
</div>

<div class="community">
<h2 data-toc-skip>Community</h2>
<ul class="list-unstyled">
<li><a href="CONTRIBUTING.html">Contributing guide</a></li>
</ul>
</div>

<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing GeoTcgaData</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>Erqiang Hu <br><small class="roles"> Author, maintainer </small> <a href="https://orcid.org/0000-0002-1798-7513" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
</ul>
</div>



  </div>
</div>


      <footer><div class="copyright">
  <p></p>
<p>Developed by Erqiang Hu.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.9000.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
