<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Do difference analysis of RNA-seq data — diff_RNA • GeoTcgaData</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Do difference analysis of RNA-seq data — diff_RNA"><meta property="og:description" content="Do difference analysis of RNA-seq data"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">GeoTcgaData</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.1.1.993</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../articles/GeoTcgaData.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"></ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Do difference analysis of RNA-seq data</h1>
    
    <div class="hidden name"><code>diff_RNA.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Do difference analysis of RNA-seq data</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">diff_RNA</span><span class="op">(</span></span>
<span>  <span class="va">counts</span>,</span>
<span>  <span class="va">group</span>,</span>
<span>  method <span class="op">=</span> <span class="st">"limma"</span>,</span>
<span>  geneLength <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  gccontent <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  filter <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  edgeRNorm <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  adjust.method <span class="op">=</span> <span class="st">"BH"</span>,</span>
<span>  useTopconfects <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>counts</dt>
<dd><p>a dataframe or numeric matrix of raw counts data</p></dd>


<dt>group</dt>
<dd><p>sample groups</p></dd>


<dt>method</dt>
<dd><p>one of "DESeq2", "edgeR" , "limma", "dearseq" and "Wilcoxon".</p></dd>


<dt>geneLength</dt>
<dd><p>a vector of gene length.</p></dd>


<dt>gccontent</dt>
<dd><p>a vector of gene GC content.</p></dd>


<dt>filter</dt>
<dd><p>if TRUE, use filterByExpr to filter genes.</p></dd>


<dt>edgeRNorm</dt>
<dd><p>if TRUE, use edgeR to do normalization for dearseq method.</p></dd>


<dt>adjust.method</dt>
<dd><p>character string specifying the method used to adjust p-values for multiple testing.
See <a href="https://rdrr.io/r/stats/p.adjust.html" class="external-link">p.adjust</a> for possible values.</p></dd>


<dt>useTopconfects</dt>
<dd><p>if TRUE, use topconfects to provide a more biologically useful ranked gene list.</p></dd>

</dl></div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="kw">if</span> <span class="op">(</span><span class="cn">FALSE</span><span class="op">)</span> <span class="op">{</span></span></span>
<span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/BioinformaticsFMRP/TCGAbiolinks" class="external-link">TCGAbiolinks</a></span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">query</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/TCGAbiolinks/man/GDCquery.html" class="external-link">GDCquery</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>  project <span class="op">=</span> <span class="st">"TCGA-ACC"</span>,</span></span>
<span class="r-in"><span>  data.category <span class="op">=</span> <span class="st">"Transcriptome Profiling"</span>,</span></span>
<span class="r-in"><span>  data.type <span class="op">=</span> <span class="st">"Gene Expression Quantification"</span>,</span></span>
<span class="r-in"><span>  workflow.type <span class="op">=</span> <span class="st">"STAR - Counts"</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/pkg/TCGAbiolinks/man/GDCdownload.html" class="external-link">GDCdownload</a></span><span class="op">(</span><span class="va">query</span>,</span></span>
<span class="r-in"><span>  method <span class="op">=</span> <span class="st">"api"</span>, files.per.chunk <span class="op">=</span> <span class="fl">3</span>,</span></span>
<span class="r-in"><span>  directory <span class="op">=</span> <span class="va">Your_Path</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">dataRNA</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/TCGAbiolinks/man/GDCprepare.html" class="external-link">GDCprepare</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>  query <span class="op">=</span> <span class="va">query</span>, directory <span class="op">=</span> <span class="va">Your_Path</span>,</span></span>
<span class="r-in"><span>  save <span class="op">=</span> <span class="cn">TRUE</span>, save.filename <span class="op">=</span> <span class="st">"dataRNA.RData"</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="co">## get raw count matrix</span></span></span>
<span class="r-in"><span><span class="va">dataPrep</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/TCGAbiolinks/man/TCGAanalyze_Preprocessing.html" class="external-link">TCGAanalyze_Preprocessing</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>  object <span class="op">=</span> <span class="va">dataRNA</span>,</span></span>
<span class="r-in"><span>  cor.cut <span class="op">=</span> <span class="fl">0.6</span>,</span></span>
<span class="r-in"><span>  datatype <span class="op">=</span> <span class="st">"STAR - Counts"</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Use `diff_RNA` to do difference analysis.</span></span></span>
<span class="r-in"><span><span class="co"># We provide the data of human gene length and GC content in `gene_cov`.</span></span></span>
<span class="r-in"><span><span class="va">group</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/AnnotationDbi/man/Bimap-envirAPI.html" class="external-link">sample</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"grp1"</span>, <span class="st">"grp2"</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">dataPrep</span><span class="op">)</span>, replace <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">cqn</span><span class="op">)</span> <span class="co"># To avoid reporting errors: there is no function "rq"</span></span></span>
<span class="r-in"><span><span class="co">## get gene length and GC content</span></span></span>
<span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">org.Hs.eg.db</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">genes_bitr</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/clusterProfiler/man/bitr.html" class="external-link">bitr</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">gene_cov</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>  fromType <span class="op">=</span> <span class="st">"ENTREZID"</span>, toType <span class="op">=</span> <span class="st">"ENSEMBL"</span>,</span></span>
<span class="r-in"><span>  OrgDb <span class="op">=</span> <span class="va">org.Hs.eg.db</span>, drop <span class="op">=</span> <span class="cn">TRUE</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">genes_bitr</span> <span class="op">&lt;-</span> <span class="va">genes_bitr</span><span class="op">[</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/duplicated.html" class="external-link">duplicated</a></span><span class="op">(</span><span class="va">genes_bitr</span><span class="op">[</span>, <span class="fl">2</span><span class="op">]</span><span class="op">)</span>, <span class="op">]</span></span></span>
<span class="r-in"><span><span class="va">gene_cov2</span> <span class="op">&lt;-</span> <span class="va">gene_cov</span><span class="op">[</span><span class="va">genes_bitr</span><span class="op">$</span><span class="va">ENTREZID</span>, <span class="op">]</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">gene_cov2</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">genes_bitr</span><span class="op">$</span><span class="va">ENSEMBL</span></span></span>
<span class="r-in"><span><span class="va">genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sets.html" class="external-link">intersect</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">dataPrep</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">gene_cov2</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">dataPrep</span> <span class="op">&lt;-</span> <span class="va">dataPrep</span><span class="op">[</span><span class="va">genes</span>, <span class="op">]</span></span></span>
<span class="r-in"><span><span class="va">geneLength</span> <span class="op">&lt;-</span> <span class="fu">gene_cov2</span><span class="op">(</span><span class="va">genes</span>, <span class="st">"length"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">gccontent</span> <span class="op">&lt;-</span> <span class="fu">gene_cov2</span><span class="op">(</span><span class="va">genes</span>, <span class="st">"GC"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">geneLength</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">gccontent</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">genes</span></span></span>
<span class="r-in"><span><span class="co">##  Difference analysis</span></span></span>
<span class="r-in"><span><span class="va">DEGAll</span> <span class="op">&lt;-</span> <span class="fu">diff_RNA</span><span class="op">(</span></span></span>
<span class="r-in"><span>  counts <span class="op">=</span> <span class="va">dataPrep</span>, group <span class="op">=</span> <span class="va">group</span>,</span></span>
<span class="r-in"><span>  geneLength <span class="op">=</span> <span class="va">geneLength</span>, gccontent <span class="op">=</span> <span class="va">gccontent</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="co"># Use `clusterProfiler` to do enrichment analytics:</span></span></span>
<span class="r-in"><span><span class="va">diffGenes</span> <span class="op">&lt;-</span> <span class="va">DEGAll</span><span class="op">$</span><span class="va">logFC</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">diffGenes</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">DEGAll</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">diffGenes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sort.html" class="external-link">sort</a></span><span class="op">(</span><span class="va">diffGenes</span>, decreasing <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://yulab-smu.top/biomedical-knowledge-mining-book/" class="external-link">clusterProfiler</a></span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://yulab-smu.top/biomedical-knowledge-mining-book/" class="external-link">enrichplot</a></span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">org.Hs.eg.db</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">gsego</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/clusterProfiler/man/gseGO.html" class="external-link">gseGO</a></span><span class="op">(</span>gene <span class="op">=</span> <span class="va">diffGenes</span>, OrgDb <span class="op">=</span> <span class="va">org.Hs.eg.db</span>, keyType <span class="op">=</span> <span class="st">"ENSEMBL"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/pkg/enrichplot/man/dotplot.html" class="external-link">dotplot</a></span><span class="op">(</span><span class="va">gsego</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">}</span></span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Erqiang Hu.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.9000.</p>
</div>

      </footer></div>

  


  

  </body></html>

